server/mqtt.go

```diff
@@ -4241,52 +4265,85 @@ func (s *Server) mqttProcessPubRel(c *client, pi uint16, trace bool) error {
 // Invoked from the MQTT publisher's readLoop. No client lock is held on entry.
 func (c *client) mqttHandlePubRetain() {
 	pp := c.mqtt.pp
-	if !mqttIsRetained(pp.flags) {
+	isRetained := mqttIsRetained(pp.flags)
+	isSparkbBirth, typ, groupID, nodeID, deviceID := sparkbIsBirth(pp.subject, sparkbNamespacePrefix)
+	if !isRetained && !isSparkbBirth {
 		return
 	}
-	key := string(pp.subject)
+
+	// Always clear the retain flag to deliver a normal published message.
+	defer func() {
+		pp.flags &= ^mqttPubFlagRetain
+	}()
+
 	asm := c.mqtt.asm
-	// Spec [MQTT-3.3.1-11]. Payload of size 0 removes the retained message,
-	// but should still be delivered as a normal message.
-	if pp.sz == 0 {
+	rm := &mqttRetainedMsg{
+		Origin: asm.jsa.id,
+		Msg:    pp.msg, // will copy these bytes later as we process rm.
+		Flags:  pp.flags,
+		Source: c.opts.Username,
+	}
+
+	key := string(pp.subject)
+	switch {
+	// Spec [MQTT-3.3.1-11]. Payload of size 0 removes the retained message, but
+	// should still be delivered as a normal message. We do not special-case
+	// sparkB BIRTH/DEATH messages since they are neither retained nor empty.
+	case pp.sz == 0:
 		if seqToRemove := asm.handleRetainedMsgDel(key, 0); seqToRemove > 0 {
 			asm.deleteRetainedMsg(seqToRemove)
 			asm.notifyRetainedMsgDeleted(key, seqToRemove)
 		}
-	} else {
+		return
+
+	case isSparkbBirth:
+		// Spec [tck-id-conformance-mqtt-aware-nbirth-mqtt-topic]. A Sparkplug
+		// Aware MQTT Server MUST make NBIRTH messages available on a topic of
+		// the form:
+		// $sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id
+		rm.Topic = fmt.Sprintf("$sparkplug/certificates/%s/%s/%s/%s", sparkbNamespace, groupID, typ, nodeID)
+		if deviceID != _EMPTY_ {
+			rm.Topic += "/" + deviceID
+		}
+		byteSubject, _ := mqttTopicToNATSPubSubject([]byte(rm.Topic))
+
+		// will use to save the retained message.
+		key = string(byteSubject)
+
+		// Copy the payload out of pp since we will be sending the message
+		// asynchronously.
+		msg := make([]byte, pp.sz)
+		copy(msg, pp.msg[:pp.sz])
+		asm.jsa.sendMsg(key, msg)
```

Comment:
@kozlovic This will send new notification messages on `$sparkplug...` subject. They need to be delivered as QoS0 messages, so thankfully sending a NATS message is sufficient. I figured I can just leverage `jsa` for it, but it uses the system account, not 100% if this is proper. 

```
[tck-id-conformance-mqtt-aware-nbirth-mqtt-retain] A Sparkplug Aware MQTT Server MUST make
NBIRTH messages available on the topic:
$sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id with the MQTT retain flag set to
true
```

Reply:
No, the asm.jsa.sendMsg() will send using an internal client on the account the session was created for, so in other words, will use the account of the incoming published message.